### PR TITLE
make `odbcDataType()` respect integer64

### LINKED ISSUE #514: odbcDataType converts integer64 to double
### Issue Description and Expected Result
`odbcDataType` method converts integer64 to double/float. It should return integer64/bigint. This happends because `typeof(as.integer64(1))` returns double.

Should be remedied with minor changes:
```r
object_type <- function(obj) {
  if (is.factor(obj)) return("factor")
  if (is(obj, "POSIXct")) return("datetime")
  if (is(obj, "Date")) return("date")
  if (is_blob(obj)) return("binary")
  if (is(obj, "difftime")) return("time")
  if(is(obj, "integer64")) return("int64") # FIX
  
  return(typeof(obj))
}

`odbcDataType.Microsoft SQL Server` <- function(con, obj, ...) {
  switch_type(obj,
    factor = varchar(obj),
    datetime = "DATETIME",
    date = "DATE",
    time = "TIME",
    binary = varbinary(obj),
    integer = "INT",
    double = "FLOAT",
    character = varchar(obj),
    logical = "BIT",
    list = varchar(obj),
    int64 = "BIGINT", # FIX
    stop("Unsupported type", call. = FALSE)
  )
}
```

### Database
Microsoft SQL Server

### Reproducible Example
``` r
library(odbc)
library(DBI)
suppressPackageStartupMessages(library(bit64))
library(dbplyr)
odbc::odbcDataType(dbplyr::simulate_mssql(), bit64::as.integer64(1))
#> [1] "FLOAT"
```

<sup>Created on 2022-10-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details>
<summary>Session Info</summary>

```r
devtools::session_info()
─ Session info ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.1.0 (2021-05-18)
 os       Ubuntu 18.04.6 LTS          
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language en_US:en                    
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       UTC                         
 date     2022-10-18                  

─ Packages ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date       lib source        
 assertthat    0.2.1   2019-03-21 [2] RSPM (R 4.1.0)
 backports     1.1.9   2020-08-24 [1] CRAN (R 4.1.0)
 bit           4.0.4   2020-08-04 [2] RSPM (R 4.1.0)
 bit64         4.0.5   2020-08-30 [1] CRAN (R 4.1.0)
 blob          1.2.1   2020-01-20 [1] CRAN (R 4.1.0)
 callr         3.7.2   2022-08-22 [2] RSPM (R 4.1.0)
 cli           3.3.0   2022-04-25 [1] CRAN (R 4.1.0)
 codetools     0.2-16  2018-12-24 [1] CRAN (R 4.1.0)
 crayon        1.3.4   2017-09-16 [1] CRAN (R 4.1.0)
 DBI           1.1.3   2022-06-18 [1] CRAN (R 4.1.0)
 dbplyr        2.2.1   2022-06-27 [1] CRAN (R 4.1.0)
 digest        0.6.25  2020-02-23 [1] CRAN (R 4.1.0)
 dplyr         1.0.10  2022-09-01 [1] CRAN (R 4.1.0)
 ellipsis      0.3.2   2021-04-29 [2] RSPM (R 4.1.0)
 evaluate      0.17    2022-10-07 [1] CRAN (R 4.1.0)
 fansi         0.4.1   2020-01-08 [1] CRAN (R 4.1.0)
 fs            1.5.0   2020-07-31 [1] CRAN (R 4.1.0)
 generics      0.0.2   2018-11-29 [1] CRAN (R 4.1.0)
 glue          1.6.2   2022-02-24 [1] CRAN (R 4.1.0)
 highr         0.8     2019-03-20 [1] CRAN (R 4.1.0)
 hms           0.5.3   2020-01-08 [1] CRAN (R 4.1.0)
 htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.1.0)
 knitr         1.39    2022-04-26 [1] CRAN (R 4.1.0)
 lgr           0.4.1   2020-10-20 [1] CRAN (R 4.1.0)
 lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.1.0)
 magrittr      1.5     2014-11-22 [1] CRAN (R 4.1.0)
 odbc          1.2.3   2020-06-18 [1] CRAN (R 4.1.0)
 pillar        1.8.1   2022-08-19 [1] CRAN (R 4.1.0)
 pkgconfig     2.0.3   2019-09-22 [2] RSPM (R 4.1.0)
 processx      3.6.1   2022-06-17 [1] CRAN (R 4.1.0)
 pryr          0.1.4   2018-02-18 [1] CRAN (R 4.1.0)
 ps            1.3.4   2020-08-11 [1] CRAN (R 4.1.0)
 R6            2.4.1   2019-11-12 [1] CRAN (R 4.1.0)
 Rcpp          1.0.8.3 2022-03-17 [1] CRAN (R 4.1.0)
 renv          0.15.5  2022-05-26 [1] CRAN (R 4.1.0)
 reprex        2.0.2   2022-08-17 [1] CRAN (R 4.1.0)
 rlang         1.0.6   2022-09-24 [1] CRAN (R 4.1.0)
 rmarkdown     2.6     2020-12-14 [1] CRAN (R 4.1.0)
 rprojroot     1.3-2   2018-01-03 [1] CRAN (R 4.1.0)
 rstudioapi    0.11    2020-02-07 [1] CRAN (R 4.1.0)
 sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.1.0)
 sloop         1.0.1   2019-02-17 [1] CRAN (R 4.1.0)
 stringi       1.4.6   2020-02-17 [1] CRAN (R 4.1.0)
 stringr       1.4.1   2022-08-20 [2] RSPM (R 4.1.0)
 tibble        3.1.2   2021-05-16 [1] CRAN (R 4.1.0)
 tidyselect    1.2.0   2022-10-10 [1] CRAN (R 4.1.0)
 utf8          1.1.4   2018-05-24 [1] CRAN (R 4.1.0)
 vctrs         0.4.2   2022-09-29 [1] CRAN (R 4.1.0)
 withr         2.5.0   2022-03-03 [1] CRAN (R 4.1.0)
 xfun          0.31    2022-05-10 [1] CRAN (R 4.1.0)
 yaml          2.2.1   2020-02-01 [1] CRAN (R 4.1.0)
```
</details>

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #514.

### CODE REVIEW

**Comment by simonpcouch** [comment]
With this PR:

``` r
library(odbc)
library(DBI)
suppressPackageStartupMessages(library(bit64))
library(dbplyr)
odbc::odbcDataType(dbplyr::simulate_mssql(), bit64::as.integer64(1))
#> [1] "BIGINT"
```

<sup>Created on 2023-12-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Review by hadley — COMMENTED** [review]
My only worry with this change is that it's going to break use of integer64 with other backends because it will now hit the unsupported type clause.  So I think you should update all the `odbcDataType()` methods to add integer64 to the switch, just using the same value as integer.

**Comment by simonpcouch** [comment]
Totally makes sense. [628879c](https://github.com/r-dbi/odbc/pull/698/commits/628879c2666df84ff29c2b10bd49157fbc3dedeb).

File: NEWS.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 * Use correct parent class for Oracle (#685).
 
+* The Microsoft SQL Server method for `odbcDataType()` will now return `"BIGINT"`
```
Comment (hadley) [inline NEWS.md:8]: Conventionally we usually put new news bullets at the top. No need to change here, just remember in the future.

**Review by hadley — APPROVED** [review]

File: NEWS.md (hunk 1 above)
Comment (simonpcouch) [reply NEWS.md]: Heard. Had to merge NEWS upstream anyway so just bumped that entry up.🐧

**Pull request merged**